src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -33,10 +35,23 @@ object FunctionBuilder {
     new ClassReader(bytes).accept(tcv, 0)
   }
 
+  def functionBuilder[R](implicit rti: TypeInfo[R]): FunctionBuilder[AsmFunction0[R]] =
+    new FunctionBuilder(Array[MaybeGenericTypeInfo[_]](), GenericTypeInfo(rti), "is/hail/asm4s/AsmFunction0")
+
+  def functionBuilder[A,R](implicit ati: TypeInfo[A], rti: TypeInfo[R]): FunctionBuilder[AsmFunction1[A, R]] =
+    new FunctionBuilder(Array(GenericTypeInfo(ati)), GenericTypeInfo(rti), "is/hail/asm4s/AsmFunction1")
+
+  def functionBuilder[A,B,R](implicit ati: TypeInfo[A], bti: TypeInfo[B], rti: TypeInfo[R]): FunctionBuilder[AsmFunction2[A, B, R]] =
+    new FunctionBuilder(Array(GenericTypeInfo(ati), GenericTypeInfo(bti)), GenericTypeInfo(rti), "is/hail/asm4s/AsmFunction2")
+
+  private implicit def methodNodeToGrowable(mn: MethodNode): Growable[AbstractInsnNode] = new Growable[AbstractInsnNode] {
+    def +=(e: AbstractInsnNode) = { mn.instructions.add(e); this }
+    def clear() { throw new UnsupportedOperationException() }
+  }
 }
 
-abstract class FunctionBuilder[R](parameterTypeInfo: Array[TypeInfo[_]], returnTypeInfo: TypeInfo[R],
-  packageName: String = "is/hail/codegen/generated") {
+class FunctionBuilder[F >: Null](parameterTypeInfo: Array[MaybeGenericTypeInfo[_]], returnTypeInfo: MaybeGenericTypeInfo[_],
+  interfaceName: String, packageName: String = "is/hail/codegen/generated") {
```

Comment:
I spent some time trying to do this, but if I pass `F = AsmFunction[Int, Int]`, I'm not sure how to navigate the `TypeTag` API to get the `Int`s and then line them up with the type parameters on the `apply` method. I'm sure it's possible, but I can't figure it out in the time I've spent thus far. The Java side of things seems like a non-starter as well. `Class` is erased, and I don't know how to get `Type` without constructing one explicitly.